hub/graphql/types/stats.py

```diff
@@ -0,0 +1,655 @@
+    return_columns: Optional[List[str]] = None
+
+
+"""
+# Some examples of use
+
+Show me the number of reform votes in each area, grouped by reform votes
+modelling swing from conservative and lab to reform:
+
+query SwingToReformByRegion {
+  statistics(
+    # gssCodes:"E15000005",
+    areaQueryMode: AREA_OR_CHILDREN,
+    # groupByArea: european_electoral_region,
+    groupByColumns: [
+      {
+        column: "reform",
+        aggregationOperation: "count"
+      }
+    ]
+    sourceIds: [
+      "5336849b-dea5-43cd-b973-dc507e5301af",
+      # "711400aa-f9c7-439f-9912-2dbe64c3c1cd"
+  	]
+    preGroupByCalculatedColumns: [
+      {
+        name: "conservative",
+        expression: "conservative - 0.03"
+        aggregationOperation: Mean
+      },
+      {
+        name: "labour",
+        expression: "labour - 0.03"
+        aggregationOperation: Mean
+      },
+      {
+        name: "reform",
+        expression: "reform + 0.06"
+        aggregationOperation: Mean
+      }
+    ]
+    calculatedColumns:[{
+    	name:"ref_marginality",
+      expression:"reform / second"
+      # expression:"first"
+      # expression:"reform / labour"
+    }]
+    returnColumns: [
+      "first_label",
+      "second_label",
+      "first",
+      "second",
+      # "labour",
+      "majority",
+      # "reform_swing"
+      "label",
+      "area_type",
+      "gss"]
+    # aggregationOperation: Mean
+  )
+}
+"""
+
+
+def statistics(
+    conf: StatisticsConfig,
+    as_grouped_data: bool = False,
+    category_key: Optional[str] = None,
+    count_key: Optional[str] = None,
+    return_numeric_keys_only: Optional[bool] = False,
+):
+    pre_calcs = (
+        [c for c in conf.pre_group_by_calculated_columns if not c.ignore]
+        if conf.pre_group_by_calculated_columns
+        else []
+    )
+    post_calcs = (
+        [c for c in conf.calculated_columns if not c.ignore]
+        if conf.calculated_columns
+        else []
+    )
+
+    # --- Get the required data for the source ---
+    qs = models.GenericData.objects.filter(
+        data_type__data_set__external_data_source_id__in=conf.source_ids
+    )
+
+    area_type_filter = None
+    # if group_by_area:
+    #     area_type_filter = postcodeIOKeyAreaTypeLookup[group_by_area]
+
+    if conf.map_bounds:
+        # area_type_filter = postcodeIOKeyAreaTypeLookup[group_by_area]
+        bbox_coords = (
+            (conf.map_bounds.west, conf.map_bounds.north),  # Top left
+            (conf.map_bounds.east, conf.map_bounds.north),  # Top right
+            (conf.map_bounds.east, conf.map_bounds.south),  # Bottom right
+            (conf.map_bounds.west, conf.map_bounds.south),  # Bottom left
+            (
+                conf.map_bounds.west,
+                conf.map_bounds.north,
+            ),  # Back to start to close polygon
+        )
+        bbox = Polygon(bbox_coords, srid=4326)
+        # areas = models.Area.objects.filter(**area_type_filter.query_filter).filter(
```

Comment:
question: should the filter by combined area be removed? 

It was there because when "rolling up" data to larger areas, not all of the relevant points are in the `map_bounds` box. 

E.G. If have data at the `output_area` level, and I'm browsing at the `LSOA` level, not all the `output_areas` for the `LSOA` I'm looking at will be in the bounding box.